File: curso/modulo-4/aula-13/brute_force.py
import hashlib
from blake3 import blake3
from passlib.hash import argon2
from datetime import *


mensagem = b'desenvolvedorio'

print('============================ Blake3 ============================')
daqui20Segundos = datetime.now() + timedelta(seconds=20)
blake3Execucoes = 0
while datetime.now() < daqui20Segundos:
    blake3(mensagem).digest()
    blake3Execucoes += 1

print('Execuções Blake3 em 20 segundos:', blake3Execucoes)

print('============================ MD5 ============================')
daqui20Segundos = datetime.now() + timedelta(seconds=20)
md5Execucoes = 0
while datetime.now() < daqui20Segundos:
    hashlib.md5(mensagem).digest()
    md5Execucoes += 1

print('Execuções MD5 em 20 segundos:', md5Execucoes)


print('============================ SHA-256 ============================')
daqui20Segundos = datetime.now() + timedelta(seconds=20)
md5Execucoes = 0
while datetime.now() < daqui20Segundos:
    hashlib.sha256(mensagem).digest()
    md5Execucoes += 1

print('Execuções SHA-256 em 20 segundos:', md5Execucoes)

print('============================ Argon2 ============================')
daqui20Segundos = datetime.now() + timedelta(seconds=20)
md5Execucoes = 0
while datetime.now() < daqui20Segundos:
    argon2.using(memory_cost=131072,rounds=6).hash(mensagem)
    md5Execucoes += 1

print('Execuções Argon2 em 20 segundos:', md5Execucoes)

# Blake3: 18.158.227
# MD5: 33.198.875
# SHA-256: 30.124.505
# Argon2: 106

File: curso/modulo-4/aula-13/brute_force.cs
#r "nuget: Blake3, 0.5.0"
#r "nuget: Sodium.Core, 1.2.3"

using System.Security.Cryptography;
using System.Text;
using Sodium;

var MD5 = (string password, out double totaltime, out int count) =>
{
    var seconds = Stopwatch.StartNew();
    var hashCount = 0;
    var bytes = Encoding.ASCII.GetBytes(password);
    var md5 = System.Security.Cryptography.MD5.Create();
    do
    {
        md5.ComputeHash(bytes);
        hashCount++;
    } while (seconds.ElapsedMilliseconds < _time);

    seconds.Stop();
    totaltime = seconds.Elapsed.TotalSeconds;
    count = hashCount;
};

var Blake = (string password, out double totaltime, out int count) =>
{
    var seconds = Stopwatch.StartNew();
    var hashCount = 0;
    var bytes = Encoding.ASCII.GetBytes(password);
    
    do
    {
        Blake3.Hasher.Hash(bytes);
        hashCount++;
    } while (seconds.ElapsedMilliseconds < _time);

    seconds.Stop();
    totaltime = seconds.Elapsed.TotalSeconds;
    count = hashCount;
};

var Sha256 = (string password, out double totaltime, out int count) =>
{
    var shA256Managed = SHA256.Create();
    var bytes = Encoding.ASCII.GetBytes(password);
    
    var seconds = Stopwatch.StartNew();
    var hashCount = 0;
    do
    {
        shA256Managed.ComputeHash(bytes);
        hashCount++;
    } while (seconds.ElapsedMilliseconds < _time);

    seconds.Stop();
    totaltime = seconds.Elapsed.TotalSeconds;
    count = hashCount;
};

var Argon2 = (string password, out double totaltime, out int count) =>
{
    var seconds = Stopwatch.StartNew();
    var hashCount = 0;
    do
    {
        PasswordHash.ArgonHashString(password, PasswordHash.StrengthArgon.Moderate);
        hashCount++;
    } while (seconds.ElapsedMilliseconds < _time);

    seconds.Stop();
    totaltime = seconds.Elapsed.TotalSeconds;
    count = hashCount;
};

long _time = 20_000;
double seconds = 0;
var hashCount = 0;

var password = "Sup3rSecr3t";
Console.WriteLine($"Quantas senhas podem ser hasheadas em 20 segundos?");
Console.WriteLine();
Console.WriteLine("========================================");
Console.WriteLine("            MD5 hashes");
MD5(password, out seconds, out hashCount);
Console.WriteLine($"MD5: {hashCount:N}");

Console.WriteLine();
Console.WriteLine("========================================");
Console.WriteLine("            SHA256 hashes");

Sha256(password, out seconds, out hashCount);
Console.WriteLine($"SHA256: {hashCount:N}");

Console.WriteLine();
Console.WriteLine("========================================");
Console.WriteLine("            BLAKE hashes");

Blake(password, out seconds, out hashCount);
Console.WriteLine($"Blake3: {hashCount:N}");

Console.WriteLine();
Console.WriteLine("========================================");
Console.WriteLine("            Argon2 hashes");

Argon2(password, out seconds, out hashCount);
Console.WriteLine($"Argon2: {hashCount:N}");